src/main/java/org/broadinstitute/hellbender/tools/spark/transforms/markduplicates/MarkDuplicatesSpark.java

```diff
@@ -103,6 +129,21 @@
         });
     }
 
+    /**
+     * Sort reads into queryname order if they are not already sorted
+     */
+    protected static JavaRDD<GATKRead> querynameSortReadsIfNecessary(JavaRDD<GATKRead> reads, int numReducers, SAMFileHeader headerForTool) {
+        JavaRDD<GATKRead> sortedReadsForMarking;
+        if (ReadUtils.isReadNameGroupedBam(headerForTool)) {
+            sortedReadsForMarking = reads;
+        } else {
+            headerForTool.setSortOrder(SAMFileHeader.SortOrder.queryname);
+            JavaRDD<GATKRead> sortedReads = SparkUtils.querynameSortReads(reads, numReducers);
+            sortedReadsForMarking = ReadsSparkSource.putPairsInSamePartition(headerForTool, sortedReads, new JavaSparkContext(reads.context()));
```

Comment:
@jamesemery I would use JavaSparkContext.fromSparkContext instead of new JavaSparkContext.  I think it's the same, but it might change in future sparks.